File: py/packages/genkit/src/genkit/ai/_genkit.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Genkit API module for managing AI flows.

This module provides a wrapper around the different implementations of the
Genkit API, automatically selecting the appropriate implementation based on
whether the code is running synchronously or asynchronously.
"""

import inspect
from collections.abc import Callable
from functools import wraps
from typing import Any, TypeVar, cast

import structlog

from genkit.aio import GenkitAsync
from genkit.core.action import ActionKind
from genkit.core.registry import Registry
from genkit.sync import GenkitSync

F = TypeVar('F', bound=Callable[..., Any])


logger = structlog.get_logger()


class GenkitExperimental:
    """Veneer class for the Genkit API.

    This class serves as a factory for both synchronous and asynchronous
    implementations of Genkit. The appropriate implementation is selected based
    on the context in which it is used when the flow is decorated.

    Should one need to access the underlying synchronous or asynchronous
    implementation, the following properties can be used:

    - `ai`: Returns the asynchronous implementation of Genkit.
    - `sync`: Returns the synchronous implementation of Genkit.
    """

    def __init__(self) -> None:
        """Initializes Genkit."""
        self._registry = Registry()
        self._async_impl = GenkitAsync()
        self._sync_impl = GenkitSync()

    @property
    def ai(self) -> GenkitAsync:
        """Returns the asynchronous implementation of Genkit."""
        return self._async_impl

    @property
    def sync(self) -> GenkitSync:
        """Returns the synchronous implementation of Genkit."""
        return self._sync_impl

    def flow(self, name: str | None = None) -> Callable[[F], F]:
        """Returns a decorator that can be used to create a Genkit flow.

        This decorator automatically detects whether the decorated function is a
        coroutine function and injects the appropriate Genkit implementation as
        the first positional argument ('ai') to the function.

        Returns:
            A decorator function that wraps the provided flow function.
        """

        def decorator(func: F) -> F:
            """Decorator to wrap a flow function with appropriate context.

            Args:
                func: The flow function to be decorated.

            Returns:
                A wrapped function that injects the appropriate Genkit
                implementation.
            """
            flow_name = name if name is not None else func.__name__
            # action = self._registry.register_action(
            #    name=flow_name,
            #    kind=ActionKind.FLOW,
            #    fn=func,
            #    span_metadata={'genkit:metadata:flow:name': flow_name},
            # )
            # logger.debug('Registered action', action=action.name)

            if inspect.iscoroutinefunction(func):
                logger.debug('Registering async flow', name=flow_name)

                @wraps(func)
                async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                    """Wrapper for async functions."""
                    result = func(self._async_impl, *args, **kwargs)
                    return await result

                return cast(F, async_wrapper)
            else:
                logger.debug('Registering sync flow', name=flow_name)

                @wraps(func)
                def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
                    """Wrapper for sync functions."""
                    return func(self._sync_impl, *args, **kwargs)

                return cast(F, sync_wrapper)

        return decorator

File: go/plugins/milvus/genkit.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package milvus

import (
	"context"
	"errors"
	"sync"

	"github.com/firebase/genkit/go/ai"
	"github.com/firebase/genkit/go/core/api"
	"github.com/firebase/genkit/go/genkit"
)

type Milvus struct {
	mu      sync.Mutex    // Mutex to control access.
	initted bool          // Whether the plugin has been initialized.
	Engine  *MilvusEngine // Client for the Milvus database.
}

// Name returns the name of the plugin.
func (m *Milvus) Name() string {
	return provider
}

// Init initializes the Milvus plugin.
func (m *Milvus) Init(ctx context.Context) []api.Action {
	m.mu.Lock()
	defer m.mu.Unlock()

	if m.initted {
		panic("milvus.Init: plugin already initialized")
	}

	if m.Engine == nil {
		panic("milvus.Init: Engine not set")
	}

	if m.Engine.client == nil {
		panic("milvus.Init: client not set")
	}

	m.initted = true
	return []api.Action{}
}

type CollectionConfig struct {
	// Name is the Milvus collection name.
	Name string
	// IdKey is the column name that stores a document identifier.
	IdKey string
	// ScoreKey is the metadata key used to store similarity scores in results.
	ScoreKey string
	// VectorKey is the column name that stores the embedding vector.
	VectorKey string
	// TextKey is the column name that stores the original document text.
	TextKey string
	// VectorDim is the dimensionality of the embedding vectors.
	VectorDim int
	// Embedder is the embedding model used to embed queries and documents.
	Embedder ai.Embedder
	// EmbedderOptions are passed to the embedder when generating embeddings.
	EmbedderOptions any
}

// DefineRetriever defines a Retriever with the given configuration.
func DefineRetriever(_ context.Context, g *genkit.Genkit, cfg *CollectionConfig, opts *ai.RetrieverOptions) (*DocStore, ai.Retriever, error) {
	m, ok := genkit.LookupPlugin(g, provider).(*Milvus)
	if !ok {
		return nil, nil, errors.New("milvus plugin not found; did you call genkit.Init with the milvus plugin")
	}

	ds, err := m.newDocStore(cfg)
	if err != nil {
		return nil, nil, err
	}
	return ds, genkit.DefineRetriever(g, api.NewName(provider, cfg.Name), opts, ds.Retrieve), nil
}
